### PR TITLE
Changes the plant reagent formula and makes plant potency, plant reagent % and plant volume interact sanely, unnerfed version.

### PR DESCRIPTION
**Pull request by Shadark**
## About The Pull Request

Closes #51723 
, even if I'm completely based on that proposal.
Refactors the calculation for the ratio of reagents and their volume based on the original code and the PR by @ArcaneMusic .

_Arcane's Proposal (if I understood it correctly):_

- If the total ratio of reagents exceeds 1, calculate the proportional ratio respective to the total ratio (in order to make the sum of all ratios become 1).
- Obtain the volume of the reagent with:
`max(1, round((grown_edible.provide_volume())*(potency/100) * reagents_add[rid] * reagent_overflow_mod, 1))` - This translates to: maximum volume * potency ratio * reagent ratio * overflow modifier - I think in fact this formula has a mistake since you should exchange the original reagent ratio for the modifier, because like this:
    - For a plant with maximum volume of 50, 100 potency, 2 reagents with 0.5 and 0.6 ratio
        - reagent_overflow_mod for the 0.6 ratio = 0.6/1.1 -> 0.5454545
        - 50 * (100/100) * 0.6 * 0.5454545 = 16,363635
    - That's less than the 60% both of the original 100 potency and the new 50 capacity.

As I think most people saw, the problem with that proposal is that it considers the maximum volume for the ratio calculations instead of using the potency (as the original code did).

_My proposal_
- If the total ratio of reagents exceeds 1, calculate the proportional ratio respective to the total ratio (in order to make the sum of all ratios become 1). -> No changes from Arcane's, this solves the problem of having multiple ratios exceeding 100%.
- If the total volume of reagents exceeds the maximum capacity volume, calculate the proportion of the volume with the maximum capacity instead of potency.
    - If we have a plant with 100 potency and some reagent with a ratio of 0.1, we end with 10 units of the reagent since it's not exceeding the maximum capacity.
    - If we have multiple reagents in the same plant, exceeding capacity (but not ratio), like 0.3 and 0.3 (would result in 30+30 units), calculate the proportional ratio and use it with the maximum volume -> 0.3/0.6 -> 0.5 * 50 max capacity = 25 units.
    - If we have both exceeding ratios and capacity, do the same calculations, everything should add up to 1 (ratios) and maximum capacity.

## Why It's Good For The Game

Fixes the issue with exceeding 100% of production ratios and not being able to express all reagents in the product when exceeding maximum capacity, without changing the default behavior when not exceeding the maximum capacity.

P.S: Not sure what's the etiquette for making a PR based on another PR when naming the changelog, if it's needed to mention @ArcaneMusic feel free to say so and I'll amend the changelog.
P.S.S: I'm not adding the tag balance since this doesn't balance anything new, still having vanilla behavior when not exceeding the maximum capacity.

## Changelog
:cl: Shadark & ArcaneMusic
tweak: Plants with over 100% reagent genes now split up their reagents between all available chemicals.
fix: Plants now correctly apply percentages of reagents to a plant's overall reagent capacity.
/:cl:

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
If I tell you a glass can hold 50 mLs, how many mLs would there be if I added 10% water? If you answered 7 because the glass of water has a strength stat of 70 you'd be correct!

If I tell you a glass can hold 50 mLs, how many mLs each would there be if I added 30% water and 30% cola with a strength stat of 100? If you answered 25 for both because it would go over 50mLs so then we look at the ratio and fill it to that point, you'd be correct!!!!

absolutely stupid-confusing all for the sake of soul.

**Comment by ATH1909** [comment]
To be clear, this PR follows what I suggested here, right?
https://user-images.githubusercontent.com/6681708/84995442-b0373a00-b154-11ea-8f0f-8566c533353f.png

**Comment by ATH1909** [comment]
But yeah, this is the much, much better way to do it.

File: code/modules/hydroponics/grown.dm [hunk 1]
```diff
@@ -73,10 +73,15 @@
...
+			var/chem_cap = 0
 			for(var/reagent_id in seed.reagents_add)
 				var/datum/reagent/R  = GLOB.chemical_reagents_list[reagent_id]
 				var/amt = reagents.get_reagent_amount(reagent_id)
-				reag_txt += "\n<span class='info'>- [R.name]: [amt]</span>"
+				chem_cap += seed.reagents_add[reagent_id]
+				if(chem_cap <= 1)
+					reag_txt += "\n<span class='info'>- [R.name]: [amt]</span>"
+				else
+					reag_txt += "\n<span class='warning'>- [R.name]: [amt] (OVER 100% CAPACITY)</span>"
```
Comment (ATH1909) [inline code/modules/hydroponics/grown.dm:84]: why is this here? the number "100%" is arbitrary; a 50% acid production+10% nutriment production will hit the chem cap for a 100 potency, 50u capacity plant, and 50%+10%=60%<100%

just delete this bit, it's gonna confuse people

File: code/modules/hydroponics/seeds.dm [hunk 2]
```diff
@@ -221,20 +221,28 @@
...
+		var/obj/item/reagent_containers/food/snacks/grown/grown_edible = T
+		var/edible_max_volume = grown_edible.provide_volume()
+		var/reagent_max_ratio = 0  // Total ratio of chemicals produced
+		var/reagent_max_volume = 0  // Initial calculation of the maximum volume of chemicals produced
+		for(var/rid in reagents_add)
+			reagent_max_ratio += reagents_add[rid]
+			reagent_max_volume += reagents_add[rid] * potency
+		for(var/rid in reagents_add)
+			var/reagent_ratio = reagents_add[rid] // Ratio used for calculations, initial behavior is to consider the original ratio
+			if((reagent_max_ratio > 1) || (reagent_max_volume > edible_max_volume)) // If we're exceeding a total ratio of 1 or we're exceeding the maximum available volume
```
Comment (ATH1909) [inline code/modules/hydroponics/seeds.dm:234]: again, why do we care if reagent_max_ratio is greater than 1? that's irrelevant

File: code/modules/hydroponics/seeds.dm [hunk 3]
```diff
@@ -221,20 +221,28 @@
...
-
-		T.reagents.add_reagent(rid, amount, data)
+	if(istype(T, /obj/item/reagent_containers/food/snacks/grown))
+		var/obj/item/reagent_containers/food/snacks/grown/grown_edible = T
+		var/edible_max_volume = grown_edible.provide_volume()
+		var/reagent_max_ratio = 0  // Total ratio of chemicals produced
+		var/reagent_max_volume = 0  // Initial calculation of the maximum volume of chemicals produced
+		for(var/rid in reagents_add)
+			reagent_max_ratio += reagents_add[rid]
+			reagent_max_volume += reagents_add[rid] * potency
```
Comment (ATH1909) [inline code/modules/hydroponics/seeds.dm:231]:
```suggestion
			reagent_max_volume += max(1, round(potency * reagents_add[rid], 1))

```

**Comment by ATH1909** [comment]
goddamnit, I'm gonna have to get off of my webeditting ass and do this myself, aren't I?

File: code/modules/hydroponics/grown.dm [hunk 4]
```diff
@@ -164,6 +169,11 @@
 		reagents.del_reagent(/datum/reagent/consumable/nutriment)
 		reagents.del_reagent(/datum/reagent/consumable/nutriment/vitamin)
 
+/obj/item/reagent_containers/food/snacks/grown/proc/provide_volume()
```
Comment (ATH1909) [inline code/modules/hydroponics/grown.dm:172]: does this account for densified chemicals?

File: code/modules/hydroponics/seeds.dm (hunk 3 above)
Comment (ATH1909) [inline code/modules/hydroponics/seeds.dm:231]:
```suggestion
			reagent_max_volume += max(1, round(potency * reagents_add[rid], 1))

```

File: code/modules/hydroponics/seeds.dm [hunk 5]
```diff
@@ -221,20 +221,28 @@
...
-		if(rid == /datum/reagent/blood) // Hack to make blood in plants always O-
-			data = list("blood_type" = "O-")
-		if(rid == /datum/reagent/consumable/nutriment || rid == /datum/reagent/consumable/nutriment/vitamin)
-			// apple tastes of apple.
-			if(istype(T, /obj/item/reagent_containers/food/snacks/grown))
-				var/obj/item/reagent_containers/food/snacks/grown/grown_edible = T
-				data = grown_edible.tastes
-
-		T.reagents.add_reagent(rid, amount, data)
+	if(istype(T, /obj/item/reagent_containers/food/snacks/grown))
```
Comment (ATH1909) [inline code/modules/hydroponics/seeds.dm:224]: what if it's a grown inedible?

File: code/modules/hydroponics/grown.dm (hunk 4 above)
Comment (ATH1909) [reply code/modules/hydroponics/grown.dm:172]: actually, I think it does

**Comment by Shadark** [comment]
Since it seems this isn't going to be the chosen proposal I'm closing early so as to avoid cluttering the open PR section, in any case thanks for everyone's input.

I'm hoping for Arcane's propsal + increased capacity (ATH's proposal) as a middle ground to satisfy everyone and unclusterfuck this issue.
